x-pack/plugins/fleet/server/services/epm/packages/install.ts

```diff
@@ -841,6 +853,165 @@ export async function ensurePackagesCompletedInstall(
   return installingPackages;
 }
 
+export async function installIndexTemplatesAndPipelines({
+  installedPkg,
+  paths,
+  packageInfo,
+  esReferences,
+  savedObjectsClient,
+  esClient,
+  logger,
+  onlyForDataStreams,
+}: {
+  installedPkg?: Installation;
+  paths: string[];
+  packageInfo: PackageInfo | InstallablePackage;
+  esReferences: EsAssetReference[];
+  savedObjectsClient: SavedObjectsClientContract;
+  esClient: ElasticsearchClient;
+  logger: Logger;
+  onlyForDataStreams?: RegistryDataStream[];
+}) {
+  /**
+   * In order to install assets in parallel, we need to split the preparation step from the installation step. This
+   * allows us to know which asset references are going to be installed so that we can save them on the packages
+   * SO before installation begins. In the case of a failure during installing any individual asset, we'll have the
+   * references necessary to remove any assets in that were successfully installed during the rollback phase.
+   *
+   * This split of prepare/install could be extended to all asset types. Besides performance, it also allows us to
+   * more easily write unit tests against the asset generation code without needing to mock ES responses.
+   */
+  const experimentalDataStreamFeatures = installedPkg?.experimental_data_stream_features ?? [];
+
+  const preparedIngestPipelines = prepareToInstallPipelines(packageInfo, paths, onlyForDataStreams);
+  const preparedIndexTemplates = prepareToInstallTemplates(
+    packageInfo,
+    paths,
+    esReferences,
+    experimentalDataStreamFeatures,
+    onlyForDataStreams
+  );
+
+  // Update the references for the templates and ingest pipelines together. Need to be done together to avoid race
+  // conditions on updating the installed_es field at the same time
```

Comment:
it is possible to have a race condition if we create two custom input package policy with different dataset at the same time?

Reply:
Ah yes I forgot to call this out, I have a PR waiting to open when this is merged as I thought this PR was large enough. It addresses the race condition:

https://github.com/elastic/kibana/pull/148883